packages/server-wallet/src/chain-service/chain-service.ts

```diff
@@ -255,110 +298,140 @@ export class ChainService implements ChainServiceInterface {
-
-      const subscription = concat<ContractEvent>(
-        currentHolding,
-        obs.pipe(filter(event => event.channelId === channelId))
-      ).subscribe({
-        next: async event => {
-          switch (event.type) {
-            case Deposited:
-              this.logger.debug(
-                {channelId, tx: event.ethersEvent?.transactionHash},
-                'Observed Deposited event on-chain; beginning to wait for confirmations'
-              );
-              await this.waitForConfirmations(event.ethersEvent);
-              subscriber.holdingUpdated(event);
-              break;
-            case AssetTransferred:
-              this.logger.debug(
-                {channelId, tx: event.ethersEvent?.transactionHash},
-                'Observed AssetTransferred event on-chain; beginning to wait for confirmations'
-              );
-              await this.waitForConfirmations(event.ethersEvent);
-              subscriber.assetTransferred(event);
-              break;
-            default:
-              throw new Error('Unexpected event from contract observable');
-          }
-        },
-      });
-      const subscriptions = this.channelToSubscription.get(channelId) ?? [];
-      this.channelToSubscription.set(channelId, [...subscriptions, subscription]);
+      // Fetch the current contract holding, and emit as an event
+      this.getInitialHoldings(contract, channelId).then(initialHoldings =>
+        subscriber.holdingUpdated(initialHoldings)
+      );
     });
   }
 
-  /** Implementation note:
-   *  The following is a simplified API that assumes a single registerChannel call per channel.
-   *  If we would like to allow for multiple registrations per channel, registerChannel should return a registration ID.
-   *  unregisterChannel will require the registration ID as a parameter.
-   */
   unregisterChannel(channelId: Bytes32): void {
-    const subscriptions = this.channelToSubscription.get(channelId);
-    if (subscriptions?.length !== 1) {
-      throw new Error(
-        'Unregister channel implementation only works when there is one subscriber per channel'
-      );
+    this.channelToSubscribers.set(channelId, []);
+  }
+
+  private async whenFinalized(channelId: Bytes32, finalizesAtS: number, tryNum: number) {
+    if (tryNum > 4) return;
+
+    const nowMs = Date.now();
+    const finalizesAtMs = finalizesAtS * 1_000;
+    if (nowMs < finalizesAtMs) {
+      setTimeout(() => this.whenFinalized(channelId, finalizesAtS, 0), nowMs - finalizesAtMs);
+      return;
     }
-    subscriptions.map(s => s.unsubscribe());
+    const [, storageFinalizesAtS] = await this.nitroAdjudicator.getChannelStorage(channelId);
+    const blockTimeS = (await this.provider.getBlock(await this.provider.getBlockNumber()))
+      .timestamp;
+    if (blockTimeS >= storageFinalizesAtS) {
+      // Should we wait for 6 blocks before emitting the finalized event?
```

Comment:
It seems like this is a general purpose configuration option on the `ChainService` that should be respected whenever a chain re-org occurring is important to detect before making any irreversible decisions. So, IMO, we should use `this.blockConfirmations` in situations like this one.

Reply:
In general, I agree with you. But what is an example of an irreversible decision that would be affected by a chain reorg, in this case?

Reply:
Maybe we should have 2 chain service notifications: channel finalized and channel finalized confirmed (that is fired after 6 blocks)?

It seems to me that it is beneficial for the wallet to get a notification as soon as possible that a channel has been finalized.